src/viam/sdk/common/client_helper.hpp

```diff
@@ -0,0 +1,82 @@
+#pragma once
+
+#include <grpcpp/client_context.h>
+#include <viam/sdk/common/proto_type.hpp>
+#include <viam/sdk/common/utils.hpp>
+
+namespace viam {
+namespace sdk {
+
+template <typename ClientType, typename StubType, typename RequestType, typename ResponseType>
+class ClientHelper {
+    static void default_rsc_(RequestType&) {}
+    static void default_rhc_(const ResponseType&) {}
+    static void default_ehc_(const ::grpc::Status& status) {
+        throw std::runtime_error(status.error_message());
+    }
+
+   public:
+    using PFn = ::grpc::Status (StubType::*)(::grpc::ClientContext*,
+                                             const RequestType&,
+                                             ResponseType*);
+    explicit ClientHelper(ClientType* client, StubType* stub, PFn pfn)
+        : client_(client), stub_(stub), pfn_(pfn) {}
+
+    ClientHelper& with(const AttributeMap& extra) {
+        return with(extra, default_rsc_);
+    }
+
+    template <typename RequestSetupCallable>
+    ClientHelper& with(RequestSetupCallable&& rsc) {
+        std::forward<RequestSetupCallable>(rsc)(request_);
+        return *this;
+    }
+
+    template <typename RequestSetupCallable>
+    ClientHelper& with(const AttributeMap& extra, RequestSetupCallable&& rsc) {
+        *request_.mutable_extra() = map_to_struct(extra);
+        return with(std::forward<RequestSetupCallable>(rsc));
+    }
+
+    template <typename ResponseHandlerCallable = decltype(default_rhc_)>
+    auto invoke(ResponseHandlerCallable&& rhc = default_rhc_) {
+        return invoke(std::forward<ResponseHandlerCallable>(rhc), default_ehc_);
+    }
+
+    template <typename ResponseHandlerCallable, typename ErrorHandlerCallable>
+    auto invoke(ResponseHandlerCallable&& rhc, ErrorHandlerCallable&& ehc) {
+        *request_.mutable_name() = client_->name();
+        ::grpc::ClientContext ctx;
+        set_client_ctx_authority(ctx);
+
+        const auto result = (stub_->*pfn_)(&ctx, request_, &response_);
+        if (result.ok()) {
+            return std::forward<ResponseHandlerCallable>(rhc)(
+                const_cast<const ResponseType&>(response_));
+        }
+
+        std::forward<ErrorHandlerCallable>(ehc)(result);
+        std::abort();
```

Comment:
The `ErrorHandlerCallable` is supposed to throw. But, it might not!

In that case we could either terminate, as I'm doing now, or we could call `default_ehc_` for you. I chose the former because the latter felt like papering over a programmer error.

Definitely interested in others thoughts here.

Reply:
What if an invocation of the `ClientHelper` had an `ehc` that swallowed an error and logged it or something? Should we just return some nullish value here instead of aborting? To be clear, I agree that we shouldn't call `default_ehc_` in the absence of a throw, but I'm not convinced we should `abort` either.

Reply:
Unfortunately, I don't think that will work out. If we change this code to `return std::forward<ErrorHandlerCallable(ehc)>(result)`, then *every* ehc will need to have the same return type as `rhc` or this function won't compile. Which would mean we couldn't have a default for `ehc` since every `invoke` handler has its own return type.

So, we can't return, so we must throw, and the `ehc` didn't do that, so either we throw or we abort.

Reply:
Ah gotcha. I was thinking about something like

```
std::forward<ErrorHandlerCallable>(ehc)(result);
return ["empty"-value-of-whatever-rhc-returns];
```

but yeah seems like that's not possible given there's no template type for the rhc return value. I can just recall MongoDB Go driver methods that would swallow certain errors from the server, and it seems like using `ClientHelper` means if a non-ok gRPC status is returned, some error _must_ be thrown or the program will abort. If we _really_ wanted to swallow an error though, we could just not use `ClientHelper`.

Reply:
Given that we can't reasonably return anything here, I agree that `abort` makes the most sense. I wonder if there's some value in logging the error message ourselves, just to provide more info to the user?

Reply:
Sure, that's easy enough.